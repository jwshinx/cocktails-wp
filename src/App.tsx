import React, { useState, useEffect } from 'react'
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'
import 'bootstrap/dist/css/bootstrap.min.css'
import 'bootstrap/dist/js/bootstrap.bundle.min.js'
import _ from 'lodash'

import styles from './App.module.css'
// import IMAGE from './gsw.png'
// import LOGO from './logo.svg'

import Title from './components/title/Title'
import cocktails from './api/cocktails'
import Segment from './components/segment/Segment'
// import { Counter } from './components/Counter'
// import Footer from './components/Footer'
import { SearchInput } from './components/ui/SearchInput'
import Sorter from './components/ui/Sorter'
import Filterer from './components/ui/Filterer'
import CocktailList from './components/cocktail-list/CocktailList'
import {
  GeneralDrink,
  SegmentType,
  SegmentKeyType,
  FirstLetterSegmentKey,
  PropertyType,
  CocktailFilter,
  // DetailDrink,
  ApiRawCocktailData,
} from './models/cocktail'

export const App = () => {
  const [dataResult, setDataResult] = useState<Array<GeneralDrink>>([])
  const [query, setQuery] = useState<string>('')
  const [segmentType, setSegmentType] = useState<SegmentType>('search')
  const [segmentKey, setSegmentKey] = useState<SegmentKeyType>(
    FirstLetterSegmentKey
  )
  const [segmentValue, setSegmentValue] = useState('c')
  const [sortProperty, setSortProperty] = useState<PropertyType<GeneralDrink>>({
    property: 'strDrink',
    isDescending: false,
  })
  const [filterProperties, setFilterProperties] = useState<
    Array<CocktailFilter<GeneralDrink>>
  >([])

  useEffect(() => {
    const search = async () => {
      try {
        const url = `/${segmentType}.php?${segmentKey.value}=${segmentValue}`
        // console.log('+++> cocktails url 1:', url)
        const { data }: { data: ApiRawCocktailData } = await cocktails.get(url)
        // console.log('+++> cocktails fetched data 6:', data)

        setDataResult(data.drinks as Array<GeneralDrink>)
      } catch (error) {
        console.log('+++> api error:', error)
      }
    }

    const identifier = setTimeout(() => {
      if (segmentValue !== '') {
        search()
      }
    }, 2000)

    return () => {
      clearTimeout(identifier)
    }
  }, [segmentValue, segmentType, segmentKey])

  // console.log(`+++> dataResult:`, dataResult)
  // console.log(`+++> query:`, query)

  const onCocktailFilterChange = (
    propertyType: CocktailFilter<GeneralDrink>
  ): void => {
    const propertyOnlyMatch = filterProperties.some((filterProperty) => {
      return filterProperty.property === propertyType.property
    })

    const fullMatch = filterProperties.some((filterProperty) => {
      return (
        filterProperty.property === propertyType.property &&
        filterProperty.isTruthySelected === propertyType.isTruthySelected
      )
    })

    if (fullMatch) {
      setFilterProperties(
        filterProperties.filter(
          (filterProperty) => filterProperty.property !== propertyType.property
        )
      )
    } else if (propertyOnlyMatch) {
      setFilterProperties([
        ...filterProperties.filter(
          (filterProperty) => filterProperty.property !== propertyType.property
        ),
        propertyType,
      ])
    } else {
      setFilterProperties([...filterProperties, propertyType])
    }
  }

  return (
    <Router>
      <div className="container">
        <div className="row">
          <div className={`col-4 ${styles.sidenav}`}>
            <div className="row mt-4">
              <Title text="Cocktails" />
            </div>
            <div className="row">
              <Segment
                segmentType={segmentType}
                segmentKey={segmentKey}
                segmentValue={segmentValue}
                onSegmentTypeChange={setSegmentType}
                onSegmentKeyChange={setSegmentKey}
                onSegmentValueChange={setSegmentValue}
              />
            </div>

            <div className="row mt-3">
              {_.isEmpty(dataResult) ? (
                <></>
              ) : (
                <SearchInput
                  placeholderText="Search name further"
                  setSearchQuery={(query) => {
                    setQuery(query)
                  }}
                />
              )}
            </div>
            <div className="row mt-3">
              <Sorter
                objects={dataResult}
                onSortPropertyClick={setSortProperty}
              />
            </div>
            <div className="row">
              <Filterer
                objects={dataResult}
                properties={filterProperties}
                onFilterChange={onCocktailFilterChange}
              />
            </div>
          </div>

          <div className={`col-8 ${styles.listarea}`}>
            <div className="row mt-3 mb-3">
              <Switch>
                <Route path="/">
                  <div className="col-12">
                    <div className="container">
                      <CocktailList
                        segmentType={segmentType}
                        data={dataResult}
                        query={query}
                        sortProperty={sortProperty}
                        filterProperties={filterProperties}
                      />
                    </div>
                  </div>
                </Route>
              </Switch>
            </div>
          </div>
        </div>
      </div>
    </Router>
  )
}
