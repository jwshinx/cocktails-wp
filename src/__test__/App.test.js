/**
 * @jest-environment jsdom
 */

import { render, screen } from '@testing-library/react'
import chai from 'chai'
import chaiDom from 'chai-dom'
import ReactDOM from 'react-dom'
// import userEvent from "@testing-library/user-event";
import React from 'react'
import renderer from 'react-test-renderer'

import { App } from '../App'

chai.use(chaiDom)
// const { expect } = chai

describe('App.tsx', () => {
  it('click on a button', () => {
    render(<button>Hello World</button>)
    screen.getByText('Hello World')
  })
})

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<App />, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('should xxx', () => {
  const tree = renderer.create(<App />).toJSON()
  expect(tree).toMatchSnapshot()
})
