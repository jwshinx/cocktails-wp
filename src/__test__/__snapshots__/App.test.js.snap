// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should xxx 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-4 sidenav"
    >
      <div
        className="row mt-4"
      >
        <div
          className="title"
        >
          <h2
            data-testid="title"
          >
            Cocktails
          </h2>
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="segment"
        >
          <div
            className="form-group"
          >
            <label
              htmlFor="sorter"
            >
              Enter type of search
            </label>
            <select
              className="form-select mb-3"
              id="sorter"
              onChange={[Function]}
            >
              <option
                value="search"
              >
                search
              </option>
              <option
                value="filter"
              >
                filter
              </option>
            </select>
          </div>
          <div
            className="form-group"
          >
            <label
              htmlFor="segment-key"
            >
              Enter category of search
            </label>
            <select
              className="form-select mb-3"
              id="segment-key"
              onChange={[Function]}
            >
              <option
                value="firstLetter"
              >
                First Letter
              </option>
              <option
                value="cocktailName"
              >
                Cocktail Name
              </option>
            </select>
          </div>
          <div
            className="row"
          >
            <span>
              Enter value
            </span>
            <br />
            <div
              className="form-group"
            >
              <input
                className="form-control"
                id="search"
                onChange={[Function]}
                placeholder="(letter or cocktail name)"
                type="search"
              />
              <small
                className="form-text text-muted"
                id="emailHelp"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="row mt-3"
      />
      <div
        className="row mt-3"
      />
      <div
        className="row"
      >
        <div
          className="col-12"
        />
      </div>
    </div>
    <div
      className="col-8 listarea"
    >
      <div
        className="row mt-3 mb-3"
      >
        <div
          className="col-12"
        >
          <div
            className="container"
          >
            <div
              className=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
